src/server/pfs/server/driver.go

```diff
@@ -1407,64 +1407,199 @@ func (d *driver) deleteCommit(ctx context.Context, commit *pfs.Commit) error {
-			if err := d.createBranch(ctx, branchInfo.Branch, commitInfo.ParentCommit, nil); err != nil {
+			// Update repo size
+			// TODO this is basically wrong. Other commits could share data with the
+			// commit being removed, in which case we're subtracting too much. Doing
+			// the right thing is hard, though. Hopefully nobody ends up with a
+			// negative repo size and cares too much
+			repoInfo := &pfs.RepoInfo{}
+			if err := repos.Get(commit.Repo.Name, repoInfo); err != nil {
+				return err
+			}
+			repoInfo.SizeBytes -= commitInfo.SizeBytes
+			return repos.Put(commit.Repo.Name, repoInfo)
+		}
+
+		// Delete the commit itself and subtract the size of the commit from repo
+		// size.
+		deleteCommit(commits, commit)
+
+		// Delete all of the downstream commits of 'commit'
+		for _, subvCommit := range commitInfo.Subvenance {
+			downstreamRepo := d.commits(subvCommit.Repo.Name).ReadWrite(stm)
+			deleteCommit(downstreamRepo, subvCommit)
+		}
+
+		// Remove the commits in 'deleted' from all upstream commits' subvenance
+		for _, provCommit := range commitInfo.Provenance {
+			provCI := &pfs.CommitInfo{}
+			upstreamRepo := d.commits(provCommit.Repo.Name).ReadWrite(stm)
+			if err := upstreamRepo.Get(provCommit.ID, provCI); err != nil {
+				return err
+			}
+			// Remove deleted commits from provCI.Subvenance (copy 'from' -> 'to')
+			to := 0
+			for from, c := range provCI.Subvenance {
+				if _, ok := deleted[c.ID]; ok {
+					continue
+				}
+				provCI.Subvenance[to] = provCI.Subvenance[from]
+				to++
+			}
+			provCI.Subvenance = provCI.Subvenance[:to]
+			// Write updated commitInfo back to etcd
+			if err := upstreamRepo.Put(provCommit.ID, provCI); err != nil {
 				return err
 			}
 		}
+		return nil
+	}); err != nil {
+		return err
 	}
 
-	// Delete the commit itself and subtract the size of the commit
-	// from repo size.
-	_, err = col.NewSTM(ctx, d.etcdClient, func(stm col.STM) error {
-		repos := d.repos.ReadWrite(stm)
-		repoInfo := new(pfs.RepoInfo)
-		if err := repos.Get(commit.Repo.Name, repoInfo); err != nil {
+	if deleteScratch {
+		// Delete the scratch space for this commit
+		// TODO put scratch spaces in a collection and do this in the txn above
+		prefix, err := d.scratchCommitPrefix(ctx, commit)
+		if err != nil {
+			return err
+		}
+		if _, err = d.etcdClient.Delete(ctx, prefix, etcd.WithPrefix()); err != nil {
 			return err
 		}
-		repoInfo.SizeBytes -= commitInfo.SizeBytes
-		repos.Put(commit.Repo.Name, repoInfo)
+	}
 
-		commits := d.commits(commit.Repo.Name).ReadWrite(stm)
-		return commits.Delete(commit.ID)
-	})
+	// Rewrite all commits in affected repos so that:
+	// - No commit's parent is a deleted commit.
+	// - No branch points to a deleted commit
+	// TODO this should happen in the txn as well, but can't because we have to
+	// list branches and commits in 'modifiedRepos', and you can't list in a txn
+	visited := make(map[string]struct{}) // repos we've already visited
```

Comment:
For this case I normally like to do `map[string]bool` that way you can do `if visited[repo]` which is a little more concise / clear.

Reply:
Cool, done